status-im/status-legacy
# Fixes cryptokitties loading/ordering

[comment]: # (Please replace ... with your information. Remove < and >)
[comment]: # (To auto-close issue on merge, please insert the related issue number after # i.e fixes #566)

Fixes #6839 

### Summary

This is a completion of/improvement upon the following [work](https://github.com/status-im/status-react/pull/9810). The kitty ids are grabbed in groups of 20 and sent to :load-kitties to be loaded individually. The problem of ordering is not due to any strange behavior on the part of mapv, but due to the perfectly normal behavior of a plain hash map. It doesn't care about ordering.

This was a bit tricky to diagnose because the :CK collectible hash doesn't exist at db initialization, it is created when :load-collectible-success decides to merge the first kitty into the db under the :CK symbol and sees that it doesn't exist, after which merge creates a hash-map for the non-existent key. (this is the default behavior of merge)

To fix this, I update db in the :load-kitties handler to ensure that the :CK key is initialized to an empty ordered hash map, which I set to descending order. Here I am following the cryptokitties API, which returns the most recently created kitties first. (a custom function can be supplied to change the ordering if necessary, as I have no idea what order they use to display their cats on their site)

### Testing notes
Testing this in the app was a bit tricky. I loaded some kitties manually in the repl, but it would be good for someone to check this with actual kitties on their wallet account.

##### Functional

- 1-1 chats
- public chats
- group chats
- wallet / transactions
- dapps / app browsing
- account recovery
- new account
- user profile updates
- networks
- mailservers
- fleet
- bootnodes

##### Non-functional

- battery performance
- CPU performance / speed of the app
- network consumption

### Steps to test


Open status application
Navigate to the wallet tab
Expand a wallet with more than 20 kitties
Expand the collectibles tab and select kitties
Ideally see all kitties in the proper order



status: ready
